docs/proposals/11-cluster-health.md

```diff
@@ -0,0 +1,298 @@
+- Alternatively, patch the `kube-controller-manager` `Deployment` to disable the `nodelifecycle` controller.
+
+#### ETCDReachable
+
+The `Endpoint` monitoring part of the `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) can be re-factored into a separate component that updates the conditions with the `ETCDReachable` condition type.
+
+##### Possible Actions
+
+- When this condition transitions from status `false` to `true` (i.e. `etcd` transitions from being unreachable to reachable), the dependent `pods` could be deleted but only if they are in `CrashLoopBackoff`. This is to make sure the dependent `pods` recover as fast as possible from an `etcd` down time. This could be done by the restarter part of the current `root` [command](https://github.com/gardener/dependency-watchdog/blob/master/cmd/root.go) of the [dependency-watchdog](https://github.com/gardener/dependency-watchdog) but re-factored as a separate restarter component which restarts (deletes) some targets `pods` based on some configured (via command-line argument, environment variables, config files etc.) conditions in the `Cluster` resource but only if the `pods` are in `CrashLoopBackoff`. 
+
+#### KubeAPIServerReachableInternally
+
+The `KubeAPIServerReachableInternally` condition type can be implemented (both the monitoring and the remedial action parts) similar to the [`ETCDReachable`](#etcdreachable) condition type.
+
+This condition may not be directly usable to take action.
+However, it might be useful in avoiding taking some pointless (and potentially harmful) actions.
+For example, if `KubeAPIServerReachableInternally` is `false` then we can avoid [scaling down the `kube-controller-manager`](#possible-actions) even if `KubeAPIServerReachableExternally` is `false`.
+
+#### UnhealthyNodesThresholdReached
+
+A component can monitor the fraction of the `NotReady` nodes in a zone and update the conditions with the `UnhealthyNodesThresholdReached` condition type.
+The list of affected zones could be maintained in the [`Message`](https://github.com/kubernetes/cluster-registry/blob/master/pkg/apis/clusterregistry/v1alpha1/types.go#L168).
+
+##### Possible Actions
+
+- If necessary, `kube-controller-manager` could be scaled down as described [above](#possible-actions).
+- Some resource creations (say, `pods`) could be throttled/rate-limited.
+- The `machine-controller-manager` could be either scaled down as described [above](#possible-actions) or be enhanced to avoid (or ramp down) deletion of machines from the affected zones (or `machinedeployments`).
+- Gardener can defer any disruptive updates until later.
```

Comment:
What do you mean by Gardener - gardenlet, gardener-controller-manager, extension controllers, something else?

Reply:
> What do you mean by Gardener - gardenlet, gardener-controller-manager, extension controllers, something else?

Any and all of the above whichever is best for a specific action.. Though with gardenlet, gardener-controller-manager may not have do anything.

Reply:
Okay, could you please then adapt the GEP with details about the planned component responsibilities and actions in details?

Reply:
> Okay, could you please then adapt the GEP with details about the planned component responsibilities and actions in details?

I will trim down the proposal based on the [consensus so far](https://github.com/gardener/gardener/pull/1797#issuecomment-582571692).